qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,265 @@
+from __future__ import annotations
+
+import dataclasses
+from abc import ABC, abstractmethod
+from typing import Dict, Tuple
+
+import gymnasium
+import numpy as np
+
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
+    """Holds statistics about an archive.
+
+    Attributes:
+        coverage: Proportion of cells in the archive that have an elite - always
+            in the range $[0,1]$.
+        qd_score_bound_norm: QD score, i.e. sum of objective values of all
+            elites in the archive. All the fitnesses are normalized to $[0,1]$
+            using the fitness_bounds before computing the QD score.
+        qd_score_original: QD score as originally introduced by Pugh et al.
+            (2016). All the fitnesses are subtracted by a lower bound before
+            being summed.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+        discrete_archive_metrics_calculator: The metrics functor that was used
+            to compute the metrics.
+    """
+
+    coverage: float
+    qd_score_bound_norm: float
+    qd_score_original: float
+    max_fitness: float
+    ccdf: np.ndarray
+
+    discrete_archive_metrics_calculator: DiscreteArchiveMetricsCalculator
+
+    @property
+    def ccdf_fitness_bins(self) -> np.ndarray:
+        """Returns the evaluation positions of the ."""
+        return np.linspace(
+            start=self.discrete_archive_metrics_calculator.fitness_bounds.low.item(),
+            stop=self.discrete_archive_metrics_calculator.fitness_bounds.high.item(),
+            num=self.discrete_archive_metrics_calculator.num_points_ccdf,
+            endpoint=True,
+        )
+
+
+class DiscreteArchiveMetricsCalculator(ABC):
+    def __init__(
+        self,
+        fitness_bounds: gymnasium.spaces.Box,
+        num_points_ccdf: int = 100,
+    ):
+        assert np.all(np.isfinite(fitness_bounds.low)) and np.all(
```

Comment:
We should raise `ValueError` here -- assertions are too broad.

Reply:
Agreed! 
[TODO] if we go on with the current structure